[TIMOB-13379] Implement console.time/timeEnd

ewanharris opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-13379

Implement console.time/timeEnd according to https://console.spec.whatwg.org/#time, with the optional log on `console.time('duplicate label')` implemented, and a slight deviation to log a warning when calling `console.timeEnd('does not exist')`.

iOS can do a higher precision on the timing due to the implementation being in native but I nerfed for parity, happy to un-nerf though.

#### Test code

```js
console.time('set label'); // Set custom labels
console.time('set label'); // Should log "Label set label already exists" at a warn lebel
setTimeout(() => {
    console.timeEnd('set label'); // Should log "set label: 100ms" (not exactly 100ms)
}, 100);

console.time(); // Defaults to "default" if no value passed in
setTimeout(() => {
    console.timeEnd(); // Should log "set label: 5000ms" (not exactly 5000ms)
}, 5000);
console.timeEnd('No exist') // Should log "Label No exist does not exist" at warn level
```

#### Output

```txt
iOS
[WARN]  Label set label already exists
[WARN]  Label No exist does not exist
[INFO]  set label: 106ms
[INFO]  default: 5003ms

Android
[WARN]  Label set label already exists
[WARN]  Label No exist does not exist
[INFO]  set label: 127ms
[INFO]  default: 5004ms
```

ewanharris committed:
[TIMOB-13379] Add Android implementation of time/timeEnd

[TIMOB-13379] Fix log message, lint code

ewanharris committed:
[TIMOB-13379] Add iOS implementation of time/timeEnd

ewanharris committed:
[TIMOB-13379] Add unit test for time/timeEnd

[TIMOB-13379] Fix lint issue

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.2.0 test /Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA/android/cli/commands/_buildModule.js
  554:18  warning  Found non-literal argument to RegExp Constructor  security/detect-non-literal-regexp

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA/tests/Resources/ti.android.notificationmanager.addontest.js
  13:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA/tests/Resources/ti.builtin.addontest.js
   7:1  error    Use the global form of 'use strict'                                    strict
  15:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  19:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA/tests/Resources/ti.ui.picker.addontest.js
   85:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  122:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  123:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  171:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  172:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  175:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  176:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  245:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  246:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  252:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  253:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9909-DFYFTUC7CU34BRDJ7LIDV54AHKORE2DGFVOY3AFZYFZON22253GA/tests/Resources/ti.ui.tableview.addontest.js
  145:10  warning  A space is required after '['               array-bracket-spacing
  145:27  warning  A space is required before ']'              array-bracket-spacing
  220:10  warning  A space is required after '['               array-bracket-spacing
  220:47  warning  A space is required before ']'              array-bracket-spacing
  288:10  warning  A space is required after '['               array-bracket-spacing
  288:27  warning  A space is required before ']'              array-bracket-spacing
  322:10  warning  A space is required after '['               array-bracket-spacing
  322:27  warning  A space is required before ']'              array-bracket-spacing
  332:25  warning  A space is required after '['               array-bracket-spacing
  332:65  warning  A space is required before ']'              array-bracket-spacing
  430:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  476:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  514:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  547:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  595:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  647:5   warning  Unexpected skipped mocha test               mocha/no-skipped-tests
  941:8   warning  'error' is assigned a value but never used  no-unused-vars

✖ 33 problems (1 error, 32 warnings)
  0 errors, 16 warnings potentially fixable with the `--fix` option.

Warning: Task "eslint:src" failed. Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

ewanharris commented on android/runtime/common/src/js/console.js:
```diff
@@ -44,3 +45,27 @@ exports.error = function () {
+
+exports.time = function (label) {
+	if (times[label]) {
+		exports.warn('Label ' + label + ' already exists');
```
imo, this would read nicer using template literals. Which are supported by the v8 version I believe. What're our best practices for the JS code on Android. The eslint options specify ES5

```json
  "parserOptions": {
    "ecmaVersion": 5,
    "sourceType": "script"
  }
```

garymathews replied:
I'm fine with that since it's supported.

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9909/3/artifact/dist/mobilesdk-7.2.0.v20180306051002-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

ewanharris committed:
[TIMOB-13379] Address feedback

Make label in double quotes when logging warning, add dealloc, simplify obj-c code when setting default value

hansemannn commented on iphone/Classes/TiConsole.m:
```diff
@@ -14,4 +14,37 @@ - (void)log:(NSArray *)args
+  if (!_times) {
+    _times = [[NSMutableDictionary alloc] init];
+  }
+  NSString *label = [args componentsJoinedByString:@""] ?: @"default";
```
Should a developer be able to pass an array of argumenzts, e.g. `console.time(['test', 'foo'])`? If not, change the arg-type to `id` and use `ENSURE_SINGLE_ARG(args, NSString *);` to force a string directly.

ewanharris committed:
[TIMOB-13379] Ensure single arg on iOS

ewanharris committed:
[TIMOB-13379] Remove label when timeEnd is cakked

garymathews commented on android/runtime/common/src/js/console.js:
```diff
@@ -25,6 +25,7 @@ function join(args) {
 	}).join(' ');
 }
 
+var times = {};
```
Could you move this to the top of `console.js` ?

ewanharris committed:
[TIMOB-13379] Add docs, move variable declaration

sgtcoolguy commented on apidoc/Global/console/console.yml:
```diff
@@ -70,3 +70,26 @@ methods:
+          and a warning will be logged to the console.
+    parameters:
+      - name: label
+        summary: The label to track the timer by, defaults to "default".
```
Isn't there an actual `default` property you can set here?

ewanharris replied:
Wasn't aware of that, updated to include default and optional in docs. Making a note to familiarise myself with our doc stuff too :)

sgtcoolguy commented on android/runtime/common/src/js/console.js:
```diff
@@ -44,3 +45,28 @@ exports.error = function () {
+	if (!label) {
+		label = 'default';
+	}
+	var startTime = times[label];
```
vars should be declared at top of scope, since they get hoisted up there.

sgtcoolguy replied:
Also, given that we know V8 is newer and can support ES6 natively, is there any reason we can't start making use of es6 syntax in runtime JS code?  (i.e. template strings here, use of const/let)

sgtcoolguy replied:
e.g. https://github.com/nodejs/node/blob/master/lib/console.js#L166-L184

ewanharris replied:
@sgtcoolguy  Updated to use template literals, const and a Map instead of a plain object

ewanharris replied:
FYI I filed https://jira.appcelerator.org/browse/TIMOB-25892 to hook the linting up in this folder as `grunt lint` doesn't include it currently

ewanharris committed:
[TIMOB-13379] Address review comments

'ES6-ify' code, add default and optional to documentation

sgtcoolguy committed:
Merge branch 'master' into TIMOB-13379

sgtcoolguy commented:
Windows PR: appcelerator/titanium_mobile_windows#1209

sgtcoolguy committed:
Merge branch 'master' into TIMOB-13379
